Avoid conflicting setters on builders

The example in the _Immutable POJOs_ section of the [blog post] shows using the `serializableBuilderClass` to deserialize the model POJOs. However, in many cases this doesn't work without additional annotations because there are overloaded setters with different parameter types. Since the builder seems to be mostly intended for use with the fluent methods, I would guess the setXYZ is just there to make it work with tools that support the java bean conventions. Ideally it would only have one setter that matched the return type of the corresponding getter. Example:

[blog post]: https://aws.amazon.com/blogs/developer/aws-sdk-for-java-2-0-developer-preview/

```scala
    val mapper = new ObjectMapper()
    val instance = Instance.builder()
      .instanceId("i-12345")
      .build()
    val json = mapper.writeValueAsString(instance.toBuilder)
    mapper.readValue(json, Instance.serializableBuilderClass())
```

Gives:

```
Conflicting setter definitions for property "productCodes": software.amazon.awssdk.services.ec2.model.Instance$BuilderImpl#setProductCodes(1 params) vs software.amazon.awssdk.services.ec2.model.Instance$BuilderImpl#setProductCodes(1 params)
 at [Source: (String)"{"amiLaunchIndex":null,"imageId":null,"instanceId":"i-12345","instanceType":null,"kernelId":null,"keyName":null,"launchTime":null,"monitoring":null,"placement":null,"platform":null,"privateDnsName":null,"privateIpAddress":null,"productCodes":null,"publicDnsName":null,"publicIpAddress":null,"ramdiskId":null,"state":null,"stateTransitionReason":null,"subnetId":null,"vpcId":null,"architecture":null,"blockDeviceMappings":null,"clientToken":null,"ebsOptimized":null,"enaSupport":null,"hypervisor":null"[truncated 275 chars]; line: 1, column: 1]
com.fasterxml.jackson.databind.JsonMappingException: Conflicting setter definitions for property "productCodes": software.amazon.awssdk.services.ec2.model.Instance$BuilderImpl#setProductCodes(1 params) vs software.amazon.awssdk.services.ec2.model.Instance$BuilderImpl#setProductCodes(1 params)
 at [Source: (String)"{"amiLaunchIndex":null,"imageId":null,"instanceId":"i-12345","instanceType":null,"kernelId":null,"keyName":null,"launchTime":null,"monitoring":null,"placement":null,"platform":null,"privateDnsName":null,"privateIpAddress":null,"productCodes":null,"publicDnsName":null,"publicIpAddress":null,"ramdiskId":null,"state":null,"stateTransitionReason":null,"subnetId":null,"vpcId":null,"architecture":null,"blockDeviceMappings":null,"clientToken":null,"ebsOptimized":null,"enaSupport":null,"hypervisor":null"[truncated 275 chars]; line: 1, column: 1]
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:305)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:268)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
	at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:476)
	at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer(ObjectMapper.java:3899)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3794)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2842)
Caused by: java.lang.IllegalArgumentException: Conflicting setter definitions for property "productCodes": software.amazon.awssdk.services.ec2.model.Instance$BuilderImpl#setProductCodes(1 params) vs software.amazon.awssdk.services.ec2.model.Instance$BuilderImpl#setProductCodes(1 params)
	at com.fasterxml.jackson.databind.introspect.POJOPropertyBuilder.getSetter(POJOPropertyBuilder.java:323)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.filterBeanProps(BeanDeserializerFactory.java:609)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.addBeanProps(BeanDeserializerFactory.java:499)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.buildBeanDeserializer(BeanDeserializerFactory.java:226)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.createBeanDeserializer(BeanDeserializerFactory.java:141)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2(DeserializerCache.java:403)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:349)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264)
	... 52 more
```

--- a/codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/ListSetters.java
+++ b/codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/ListSetters.java
@@ -88,11 +88,6 @@ class ListSetters extends AbstractMemberSetters {
         List<MethodSpec> beanStyle = new ArrayList<>();
 
         beanStyle.add(beanStyleCopySetter());
-        beanStyle.add(beanStyleVarargToListSetter());
-
-        if (memberModel().getEnumType() != null) {
-            beanStyle.add(beanStyleEnumVarargToListSetter());
-        }
 
         return beanStyle;
     }
@@ -128,14 +123,6 @@ class ListSetters extends AbstractMemberSetters {
                 .build();
     }
 
-    private MethodSpec beanStyleVarargToListSetter() {
-        return beanStyleSetterBuilder(ParameterSpec.builder(asArray(), fieldName()).build())
-                .varargs(true)
-                .addAnnotation(SafeVarargs.class)
-                .addCode(varargToListSetterBody())
-                .build();
-    }
-
     private MethodSpec fluentEnumVarargToListSetter(TypeName returnType) {
         return fluentSetterBuilder(ParameterSpec.builder(asArrayOfModeledEnum(), fieldName()).build(), returnType)
                 .varargs(true)
@@ -145,13 +132,6 @@ class ListSetters extends AbstractMemberSetters {
                 .build();
     }
 
-    private MethodSpec beanStyleEnumVarargToListSetter() {
-        return beanStyleSetterBuilder(ParameterSpec.builder(asArrayOfModeledEnum(), fieldName()).build())
-                .varargs(true)
-                .addAnnotation(SafeVarargs.class)
-                .addCode(enumVarargToListSetterBody())
-                .build();
-    }
 
     private CodeBlock varargToListSetterBody() {
         return CodeBlock.of("$1L($2T.asList($1L));", fieldName(), Arrays.class);

--- a/codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/NonCollectionSetters.java
+++ b/codegen/src/main/java/software/amazon/awssdk/codegen/poet/model/NonCollectionSetters.java
@@ -73,10 +73,6 @@ class NonCollectionSetters extends AbstractMemberSetters {
 
         beanStyle.add(beanStyleAssignmentSetter());
 
-        if (memberModel().getEnumType() != null) {
-            beanStyle.add(beanStyleEnumToStringSetter());
-        }
-
         return beanStyle;
     }
 
@@ -103,12 +99,6 @@ class NonCollectionSetters extends AbstractMemberSetters {
     private MethodSpec fluentEnumToStringSetter(TypeName returnType) {
         return fluentSetterBuilder(modeledParam(), returnType)
                 .addCode(enumToStringAssignmentBody().toBuilder().addStatement("return this").build())
-                .build();
-    }
-
-    private MethodSpec beanStyleEnumToStringSetter() {
-        return beanStyleSetterBuilder(modeledParam())
-                .addCode(enumToStringAssignmentBody())
                 .build();
     }
 

--- a/codegen/src/test/java/software/amazon/awssdk/codegen/poet/model/AwsModelSpecTest.java
+++ b/codegen/src/test/java/software/amazon/awssdk/codegen/poet/model/AwsModelSpecTest.java
@@ -15,15 +15,20 @@
 
 package software.amazon.awssdk.codegen.poet.model;
 
+import static java.util.stream.Collectors.toList;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static software.amazon.awssdk.codegen.poet.PoetMatchers.generatesTo;
+import static software.amazon.awssdk.utils.FunctionalUtils.invokeSafely;
 
 import java.io.File;
 import java.io.IOException;
 import java.net.URISyntaxException;
+import java.util.Collection;
 import java.util.Locale;
 import org.junit.BeforeClass;
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
 import software.amazon.awssdk.codegen.C2jModels;
 import software.amazon.awssdk.codegen.IntermediateModelBuilder;
 import software.amazon.awssdk.codegen.model.config.customization.CustomizationConfig;
@@ -32,33 +37,42 @@ import software.amazon.awssdk.codegen.model.intermediate.ShapeModel;
 import software.amazon.awssdk.codegen.model.service.ServiceModel;
 import software.amazon.awssdk.codegen.utils.ModelLoaderUtils;
 
+@RunWith(Parameterized.class)
 public class AwsModelSpecTest {
     private static File serviceModelFile;
     private static IntermediateModel intermediateModel;
+    private final ShapeModel shapeModel;
 
-    @BeforeClass
-    public static void setUp() throws URISyntaxException, IOException {
-        serviceModelFile = new File(AwsModelSpecTest.class.getResource("service-2.json").getFile());
-        File customizationConfigFile = new File(AwsModelSpecTest.class
-                .getResource("customization.config")
-                .getFile());
+    @Parameterized.Parameters(name = "{0}")
+    public static Collection<Object[]> data() {
+        invokeSafely(AwsModelSpecTest::setUp);
+        return intermediateModel.getShapes().values().stream().map(shape -> new Object[] { shape }).collect(toList());
+    }
 
-        intermediateModel = new IntermediateModelBuilder(
-                C2jModels.builder()
-                        .serviceModel(ModelLoaderUtils.loadModel(ServiceModel.class, serviceModelFile))
-                        .customizationConfig(ModelLoaderUtils.loadModel(CustomizationConfig.class, customizationConfigFile))
-                        .build())
-                .build();
+    public AwsModelSpecTest(ShapeModel shapeModel) {
+        this.shapeModel = shapeModel;
     }
 
     @Test
     public void basicGeneration() {
-        intermediateModel.getShapes().values().forEach(shape -> {
-            assertThat(new AwsServiceModel(intermediateModel, shape), generatesTo(referenceFileFor(shape)));
-        });
+        assertThat(new AwsServiceModel(intermediateModel, shapeModel), generatesTo(referenceFileForShape()));
     }
 
-    private static String referenceFileFor(ShapeModel shapeModel) {
+    private String referenceFileForShape() {
         return shapeModel.getShapeName().toLowerCase(Locale.ENGLISH) + ".java";
+    }
+
+    private static void setUp() throws URISyntaxException, IOException {
+        serviceModelFile = new File(AwsModelSpecTest.class.getResource("service-2.json").getFile());
+        File customizationConfigFile = new File(AwsModelSpecTest.class
+                                                    .getResource("customization.config")
+                                                    .getFile());
+
+        intermediateModel = new IntermediateModelBuilder(
+            C2jModels.builder()
+                     .serviceModel(ModelLoaderUtils.loadModel(ServiceModel.class, serviceModelFile))
+                     .customizationConfig(ModelLoaderUtils.loadModel(CustomizationConfig.class, customizationConfigFile))
+                     .build())
+            .build();
     }
 }

--- a/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java
+++ b/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java
@@ -16,7 +16,7 @@ import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
  */
 @Generated("software.amazon.awssdk:codegen")
 public class AllTypesRequest extends AmazonWebServiceRequest implements
-        ToCopyableBuilder<AllTypesRequest.Builder, AllTypesRequest> {
+                                                             ToCopyableBuilder<AllTypesRequest.Builder, AllTypesRequest> {
     private final String stringMember;
 
     private final Integer integerMember;
@@ -59,6 +59,8 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
 
     private final SubTypeOne polymorphicTypeWithoutSubTypes;
 
+    private final String enumType;
+
     private AllTypesRequest(BuilderImpl builder) {
         this.stringMember = builder.stringMember;
         this.integerMember = builder.integerMember;
@@ -81,6 +83,7 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
         this.recursiveStruct = builder.recursiveStruct;
         this.polymorphicTypeWithSubTypes = builder.polymorphicTypeWithSubTypes;
         this.polymorphicTypeWithoutSubTypes = builder.polymorphicTypeWithoutSubTypes;
+        this.enumType = builder.enumType;
     }
 
     /**
@@ -299,6 +302,16 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
         return polymorphicTypeWithoutSubTypes;
     }
 
+    /**
+     * Returns the value of the EnumType property for this object.
+     *
+     * @return The value of the EnumType property for this object.
+     * @see EnumType
+     */
+    public String enumType() {
+        return enumType;
+    }
+
     @Override
     public Builder toBuilder() {
         return new BuilderImpl(this);
@@ -329,7 +342,7 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
         hashCode = 31 * hashCode + ((mapOfStringToStruct() == null) ? 0 : mapOfStringToStruct().hashCode());
         hashCode = 31 * hashCode + ((timestampMember() == null) ? 0 : timestampMember().hashCode());
         hashCode = 31 * hashCode
-                + ((structWithNestedTimestampMember() == null) ? 0 : structWithNestedTimestampMember().hashCode());
+                   + ((structWithNestedTimestampMember() == null) ? 0 : structWithNestedTimestampMember().hashCode());
         hashCode = 31 * hashCode + ((blobArg() == null) ? 0 : blobArg().hashCode());
         hashCode = 31 * hashCode + ((structWithNestedBlob() == null) ? 0 : structWithNestedBlob().hashCode());
         hashCode = 31 * hashCode + ((blobMap() == null) ? 0 : blobMap().hashCode());
@@ -337,6 +350,7 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
         hashCode = 31 * hashCode + ((recursiveStruct() == null) ? 0 : recursiveStruct().hashCode());
         hashCode = 31 * hashCode + ((polymorphicTypeWithSubTypes() == null) ? 0 : polymorphicTypeWithSubTypes().hashCode());
         hashCode = 31 * hashCode + ((polymorphicTypeWithoutSubTypes() == null) ? 0 : polymorphicTypeWithoutSubTypes().hashCode());
+        hashCode = 31 * hashCode + ((enumType() == null) ? 0 : enumType().hashCode());
         return hashCode;
     }
 
@@ -434,7 +448,7 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
             return false;
         }
         if (other.structWithNestedTimestampMember() != null
-                && !other.structWithNestedTimestampMember().equals(this.structWithNestedTimestampMember())) {
+            && !other.structWithNestedTimestampMember().equals(this.structWithNestedTimestampMember())) {
             return false;
         }
         if (other.blobArg() == null ^ this.blobArg() == null) {
@@ -471,14 +485,20 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
             return false;
         }
         if (other.polymorphicTypeWithSubTypes() != null
-                && !other.polymorphicTypeWithSubTypes().equals(this.polymorphicTypeWithSubTypes())) {
+            && !other.polymorphicTypeWithSubTypes().equals(this.polymorphicTypeWithSubTypes())) {
             return false;
         }
         if (other.polymorphicTypeWithoutSubTypes() == null ^ this.polymorphicTypeWithoutSubTypes() == null) {
             return false;
         }
         if (other.polymorphicTypeWithoutSubTypes() != null
-                && !other.polymorphicTypeWithoutSubTypes().equals(this.polymorphicTypeWithoutSubTypes())) {
+            && !other.polymorphicTypeWithoutSubTypes().equals(this.polymorphicTypeWithoutSubTypes())) {
+            return false;
+        }
+        if (other.enumType() == null ^ this.enumType() == null) {
+            return false;
+        }
+        if (other.enumType() != null && !other.enumType().equals(this.enumType())) {
             return false;
         }
         return true;
@@ -550,6 +570,9 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
         }
         if (polymorphicTypeWithoutSubTypes() != null) {
             sb.append("PolymorphicTypeWithoutSubTypes: ").append(polymorphicTypeWithoutSubTypes()).append(",");
+        }
+        if (enumType() != null) {
+            sb.append("EnumType: ").append(enumType()).append(",");
         }
         sb.append("}");
         return sb.toString();
@@ -784,6 +807,28 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
          * @return Returns a reference to this object so that method calls can be chained together.
          */
         Builder polymorphicTypeWithoutSubTypes(SubTypeOne polymorphicTypeWithoutSubTypes);
+
+        /**
+         * Sets the value of the EnumType property for this object.
+         *
+         * @param enumType
+         *        The new value for the EnumType property for this object.
+         * @see EnumType
+         * @return Returns a reference to this object so that method calls can be chained together.
+         * @see EnumType
+         */
+        Builder enumType(String enumType);
+
+        /**
+         * Sets the value of the EnumType property for this object.
+         *
+         * @param enumType
+         *        The new value for the EnumType property for this object.
+         * @see EnumType
+         * @return Returns a reference to this object so that method calls can be chained together.
+         * @see EnumType
+         */
+        Builder enumType(EnumType enumType);
     }
 
     private static final class BuilderImpl implements Builder {
@@ -829,6 +874,8 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
 
         private SubTypeOne polymorphicTypeWithoutSubTypes;
 
+        private String enumType;
+
         private BuilderImpl() {
         }
 
@@ -854,6 +901,7 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
             setRecursiveStruct(model.recursiveStruct);
             setPolymorphicTypeWithSubTypes(model.polymorphicTypeWithSubTypes);
             setPolymorphicTypeWithoutSubTypes(model.polymorphicTypeWithoutSubTypes);
+            setEnumType(model.enumType);
         }
 
         public final String getStringMember() {
@@ -961,11 +1009,6 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
             this.simpleList = ListOfStringsCopier.copy(simpleList);
         }
 
-        @SafeVarargs
-        public final void setSimpleList(String... simpleList) {
-            simpleList(Arrays.asList(simpleList));
-        }
-
         public final Collection<Map<String, String>> getListOfMaps() {
             return listOfMaps;
         }
@@ -987,11 +1030,6 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
             this.listOfMaps = ListOfMapStringToStringCopier.copy(listOfMaps);
         }
 
-        @SafeVarargs
-        public final void setListOfMaps(Map<String, String>... listOfMaps) {
-            listOfMaps(Arrays.asList(listOfMaps));
-        }
-
         public final Collection<SimpleStruct> getListOfStructs() {
             return listOfStructs;
         }
@@ -1011,11 +1049,6 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
 
         public final void setListOfStructs(Collection<SimpleStruct> listOfStructs) {
             this.listOfStructs = ListOfSimpleStructsCopier.copy(listOfStructs);
-        }
-
-        @SafeVarargs
-        public final void setListOfStructs(SimpleStruct... listOfStructs) {
-            listOfStructs(Arrays.asList(listOfStructs));
         }
 
         public final Map<String, ? extends Collection<Integer>> getMapOfStringToIntegerList() {
@@ -1151,11 +1184,6 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
             this.listOfBlobs = ListOfBlobsTypeCopier.copy(listOfBlobs);
         }
 
-        @SafeVarargs
-        public final void setListOfBlobs(ByteBuffer... listOfBlobs) {
-            listOfBlobs(Arrays.asList(listOfBlobs));
-        }
-
         public final RecursiveStructType getRecursiveStruct() {
             return recursiveStruct;
         }
@@ -1198,10 +1226,29 @@ public class AllTypesRequest extends AmazonWebServiceRequest implements
             this.polymorphicTypeWithoutSubTypes = polymorphicTypeWithoutSubTypes;
         }
 
+        public final String getEnumType() {
+            return enumType;
+        }
+
+        @Override
+        public final Builder enumType(String enumType) {
+            this.enumType = enumType;
+            return this;
+        }
+
+        @Override
+        public final Builder enumType(EnumType enumType) {
+            this.enumType(enumType.toString());
+            return this;
+        }
+
+        public final void setEnumType(String enumType) {
+            this.enumType = enumType;
+        }
+
         @Override
         public AllTypesRequest build() {
             return new AllTypesRequest(this);
         }
     }
 }
-

--- a/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesresponse.java
+++ b/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesresponse.java
@@ -60,6 +60,8 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
 
     private final SubTypeOne polymorphicTypeWithoutSubTypes;
 
+    private final String enumType;
+
     private AllTypesResponse(BuilderImpl builder) {
         this.stringMember = builder.stringMember;
         this.integerMember = builder.integerMember;
@@ -82,6 +84,7 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
         this.recursiveStruct = builder.recursiveStruct;
         this.polymorphicTypeWithSubTypes = builder.polymorphicTypeWithSubTypes;
         this.polymorphicTypeWithoutSubTypes = builder.polymorphicTypeWithoutSubTypes;
+        this.enumType = builder.enumType;
     }
 
     /**
@@ -300,6 +303,16 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
         return polymorphicTypeWithoutSubTypes;
     }
 
+    /**
+     * Returns the value of the EnumType property for this object.
+     *
+     * @return The value of the EnumType property for this object.
+     * @see EnumType
+     */
+    public String enumType() {
+        return enumType;
+    }
+
     @Override
     public Builder toBuilder() {
         return new BuilderImpl(this);
@@ -338,6 +351,7 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
         hashCode = 31 * hashCode + ((recursiveStruct() == null) ? 0 : recursiveStruct().hashCode());
         hashCode = 31 * hashCode + ((polymorphicTypeWithSubTypes() == null) ? 0 : polymorphicTypeWithSubTypes().hashCode());
         hashCode = 31 * hashCode + ((polymorphicTypeWithoutSubTypes() == null) ? 0 : polymorphicTypeWithoutSubTypes().hashCode());
+        hashCode = 31 * hashCode + ((enumType() == null) ? 0 : enumType().hashCode());
         return hashCode;
     }
 
@@ -482,6 +496,12 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
                 && !other.polymorphicTypeWithoutSubTypes().equals(this.polymorphicTypeWithoutSubTypes())) {
             return false;
         }
+        if (other.enumType() == null ^ this.enumType() == null) {
+            return false;
+        }
+        if (other.enumType() != null && !other.enumType().equals(this.enumType())) {
+            return false;
+        }
         return true;
     }
 
@@ -551,6 +571,9 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
         }
         if (polymorphicTypeWithoutSubTypes() != null) {
             sb.append("PolymorphicTypeWithoutSubTypes: ").append(polymorphicTypeWithoutSubTypes()).append(",");
+        }
+        if (enumType() != null) {
+            sb.append("EnumType: ").append(enumType()).append(",");
         }
         sb.append("}");
         return sb.toString();
@@ -785,6 +808,28 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
          * @return Returns a reference to this object so that method calls can be chained together.
          */
         Builder polymorphicTypeWithoutSubTypes(SubTypeOne polymorphicTypeWithoutSubTypes);
+
+        /**
+         * Sets the value of the EnumType property for this object.
+         *
+         * @param enumType
+         *        The new value for the EnumType property for this object.
+         * @see EnumType
+         * @return Returns a reference to this object so that method calls can be chained together.
+         * @see EnumType
+         */
+        Builder enumType(String enumType);
+
+        /**
+         * Sets the value of the EnumType property for this object.
+         *
+         * @param enumType
+         *        The new value for the EnumType property for this object.
+         * @see EnumType
+         * @return Returns a reference to this object so that method calls can be chained together.
+         * @see EnumType
+         */
+        Builder enumType(EnumType enumType);
     }
 
     private static final class BuilderImpl implements Builder {
@@ -830,6 +875,8 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
 
         private SubTypeOne polymorphicTypeWithoutSubTypes;
 
+        private String enumType;
+
         private BuilderImpl() {
         }
 
@@ -855,6 +902,7 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
             setRecursiveStruct(model.recursiveStruct);
             setPolymorphicTypeWithSubTypes(model.polymorphicTypeWithSubTypes);
             setPolymorphicTypeWithoutSubTypes(model.polymorphicTypeWithoutSubTypes);
+            setEnumType(model.enumType);
         }
 
         public final String getStringMember() {
@@ -962,11 +1010,6 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
             this.simpleList = ListOfStringsCopier.copy(simpleList);
         }
 
-        @SafeVarargs
-        public final void setSimpleList(String... simpleList) {
-            simpleList(Arrays.asList(simpleList));
-        }
-
         public final Collection<Map<String, String>> getListOfMaps() {
             return listOfMaps;
         }
@@ -988,11 +1031,6 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
             this.listOfMaps = ListOfMapStringToStringCopier.copy(listOfMaps);
         }
 
-        @SafeVarargs
-        public final void setListOfMaps(Map<String, String>... listOfMaps) {
-            listOfMaps(Arrays.asList(listOfMaps));
-        }
-
         public final Collection<SimpleStruct> getListOfStructs() {
             return listOfStructs;
         }
@@ -1012,11 +1050,6 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
 
         public final void setListOfStructs(Collection<SimpleStruct> listOfStructs) {
             this.listOfStructs = ListOfSimpleStructsCopier.copy(listOfStructs);
-        }
-
-        @SafeVarargs
-        public final void setListOfStructs(SimpleStruct... listOfStructs) {
-            listOfStructs(Arrays.asList(listOfStructs));
         }
 
         public final Map<String, ? extends Collection<Integer>> getMapOfStringToIntegerList() {
@@ -1152,11 +1185,6 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
             this.listOfBlobs = ListOfBlobsTypeCopier.copy(listOfBlobs);
         }
 
-        @SafeVarargs
-        public final void setListOfBlobs(ByteBuffer... listOfBlobs) {
-            listOfBlobs(Arrays.asList(listOfBlobs));
-        }
-
         public final RecursiveStructType getRecursiveStruct() {
             return recursiveStruct;
         }
@@ -1197,6 +1225,26 @@ public class AllTypesResponse extends AmazonWebServiceResult<ResponseMetadata> i
 
         public final void setPolymorphicTypeWithoutSubTypes(SubTypeOne polymorphicTypeWithoutSubTypes) {
             this.polymorphicTypeWithoutSubTypes = polymorphicTypeWithoutSubTypes;
+        }
+
+        public final String getEnumType() {
+            return enumType;
+        }
+
+        @Override
+        public final Builder enumType(String enumType) {
+            this.enumType = enumType;
+            return this;
+        }
+
+        @Override
+        public final Builder enumType(EnumType enumType) {
+            this.enumType(enumType.toString());
+            return this;
+        }
+
+        public final void setEnumType(String enumType) {
+            this.enumType = enumType;
         }
 
         @Override

--- a/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/enumtype.java
+++ b/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/enumtype.java
@@ -1,0 +1,68 @@
+/*
+ * Copyright (c) 2017. Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.jsonprotocoltests.model;
+
+
+import javax.annotation.Generated;
+
+/**
+ */
+@Generated("software.amazon.awssdk:codegen")
+public class EnumType {
+    private EnumType(EnumType.BuilderImpl builder) {
+    }
+
+    @Override
+    public EnumType.Builder toBuilder() {
+        return new EnumType.BuilderImpl(this);
+    }
+
+    public static EnumType.Builder builder() {
+        return new EnumType.BuilderImpl();
+    }
+
+    public static Class<? extends EnumType.Builder> serializableBuilderClass() {
+        return EnumType.BuilderImpl.class;
+    }
+
+    @Override
+    public int hashCode() {
+        int hashCode = 1;
+        return hashCode;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (!(obj instanceof EnumType)) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("{");
+        sb.append("}");
+        return sb.toString();
+    }
+}

--- a/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/nestedcontainersrequest.java
+++ b/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/nestedcontainersrequest.java
@@ -222,11 +222,6 @@ public class NestedContainersRequest extends AmazonWebServiceRequest implements
             this.listOfListOfStrings = ListOfListOfStringsCopier.copy(listOfListOfStrings);
         }
 
-        @SafeVarargs
-        public final void setListOfListOfStrings(Collection<String>... listOfListOfStrings) {
-            listOfListOfStrings(Arrays.asList(listOfListOfStrings));
-        }
-
         public final Collection<? extends Collection<? extends Collection<String>>> getListOfListOfListOfStrings() {
             return listOfListOfListOfStrings;
         }
@@ -248,11 +243,6 @@ public class NestedContainersRequest extends AmazonWebServiceRequest implements
         public final void setListOfListOfListOfStrings(
                 Collection<? extends Collection<? extends Collection<String>>> listOfListOfListOfStrings) {
             this.listOfListOfListOfStrings = ListOfListOfListOfStringsCopier.copy(listOfListOfListOfStrings);
-        }
-
-        @SafeVarargs
-        public final void setListOfListOfListOfStrings(Collection<? extends Collection<String>>... listOfListOfListOfStrings) {
-            listOfListOfListOfStrings(Arrays.asList(listOfListOfListOfStrings));
         }
 
         public final Map<String, ? extends Collection<? extends Collection<String>>> getMapOfStringToListOfListOfStrings() {

--- a/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/nestedcontainersresponse.java
+++ b/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/nestedcontainersresponse.java
@@ -223,11 +223,6 @@ public class NestedContainersResponse extends AmazonWebServiceResult<ResponseMet
             this.listOfListOfStrings = ListOfListOfStringsCopier.copy(listOfListOfStrings);
         }
 
-        @SafeVarargs
-        public final void setListOfListOfStrings(Collection<String>... listOfListOfStrings) {
-            listOfListOfStrings(Arrays.asList(listOfListOfStrings));
-        }
-
         public final Collection<? extends Collection<? extends Collection<String>>> getListOfListOfListOfStrings() {
             return listOfListOfListOfStrings;
         }
@@ -249,11 +244,6 @@ public class NestedContainersResponse extends AmazonWebServiceResult<ResponseMet
         public final void setListOfListOfListOfStrings(
                 Collection<? extends Collection<? extends Collection<String>>> listOfListOfListOfStrings) {
             this.listOfListOfListOfStrings = ListOfListOfListOfStringsCopier.copy(listOfListOfListOfStrings);
-        }
-
-        @SafeVarargs
-        public final void setListOfListOfListOfStrings(Collection<? extends Collection<String>>... listOfListOfListOfStrings) {
-            listOfListOfListOfStrings(Arrays.asList(listOfListOfListOfStrings));
         }
 
         public final Map<String, ? extends Collection<? extends Collection<String>>> getMapOfStringToListOfListOfStrings() {

--- a/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/recursivestructtype.java
+++ b/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/recursivestructtype.java
@@ -276,11 +276,6 @@ public class RecursiveStructType implements StructuredPojo, ToCopyableBuilder<Re
             this.recursiveList = RecursiveListTypeCopier.copy(recursiveList);
         }
 
-        @SafeVarargs
-        public final void setRecursiveList(RecursiveStructType... recursiveList) {
-            recursiveList(Arrays.asList(recursiveList));
-        }
-
         public final Map<String, RecursiveStructType> getRecursiveMap() {
             return recursiveMap;
         }

--- a/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/service-2.json
+++ b/codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/service-2.json
@@ -82,7 +82,8 @@
         "ListOfBlobs":{"shape":"ListOfBlobsType"},
         "RecursiveStruct":{"shape":"RecursiveStructType"},
         "PolymorphicTypeWithSubTypes":{"shape":"BaseType"},
-        "PolymorphicTypeWithoutSubTypes":{"shape":"SubTypeOne"}
+        "PolymorphicTypeWithoutSubTypes":{"shape":"SubTypeOne"},
+        "EnumType":{"shape":"EnumType"}
       }
     },
     "BaseType":{
@@ -226,6 +227,12 @@
       "members":{
         "SubTypeOneMember":{"shape":"String"}
       }
+    },
+    "EnumType": {
+      "type":"string",
+      "enum": [
+        "EnumValue1", "EnumValue2"
+      ]
     },
     "Timestamp":{"type":"timestamp"}
   }

